### PR TITLE
pkg/endpoint: check for regenerateBPF error earlier

### PR DESCRIPTION
**Pull request by aanm**
It's confusing to send an error to a function and not checking it
earlier closer to the function that returned that error. Made some
changes to make the code easier for readability.

This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5377)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/endpoint/endpoint.go [hunk 1]
```diff
@@ -1233,6 +1233,12 @@ func (e *Endpoint) FailedDirectoryPath() string {
 	return filepath.Join(".", fmt.Sprintf("%d%s", e.ID, "_next_fail"))
 }
 
+// NextDirectoryPath returns the directory name for this endpoint bpf program
```
Comment (ianvernon) [inline pkg/endpoint/endpoint.go]: Please put this function in `directory.go`.

File: pkg/endpoint/policy.go [hunk 2]
```diff
@@ -731,11 +731,19 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
...
+	if err != nil {
+		failDir := e.FailedDirectoryPath()
+		scopedLog.WithFields(logrus.Fields{
+			logfields.Path: failDir,
+		}).Warn("generating BPF for endpoint failed, keeping stale directory.")
+		os.Rename(tmpDir, failDir)
+		return err
+	}
 
 	// Depending upon result of BPF regeneration (compilation executed, or
```
Comment (ianvernon) [inline pkg/endpoint/policy.go]: Update comment to reflect removal of error parameter.

File: pkg/endpoint/directory.go [hunk 3]
```diff
@@ -23,31 +23,20 @@ import (
...
-}
-
 // synchronizeDirectories moves the files related to endpoint BPF program
 // compilation to their according directories based on whether an error occurred
 // during regeneration, or if compilation of BPF was necessary for the endpoint.
 // Returns the original regenerationError if regenerationError was non-nil,
 // or if any updates to directories for the endpoint's directories fails.
 // Must be called with endpoint.Mutex held.
-func (e *Endpoint) synchronizeDirectories(origDir string, compilationExecuted bool, regenerationError error) error {
+func (e *Endpoint) synchronizeDirectories(origDir string, compilationExecuted bool) error {
```
Comment (ianvernon) [inline pkg/endpoint/directory.go:50]: Update documentation to remove references to regeneration error

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: pkg/endpoint/directory.go (hunk 3 above)
Comment (aanm) [reply pkg/endpoint/directory.go:50]: Done

File: pkg/endpoint/endpoint.go (hunk 1 above)
Comment (aanm) [reply pkg/endpoint/endpoint.go]: Done

File: pkg/endpoint/policy.go (hunk 2 above)
Comment (aanm) [reply pkg/endpoint/policy.go]: Done

**Comment by aanm** [comment]
test-me-please

File: pkg/endpoint/endpoint.go [hunk 4]
```diff
@@ -2074,8 +2080,8 @@ func (e *Endpoint) SetStateLocked(toState, reason string) bool {
 		e.getLogger().WithFields(logrus.Fields{
 			logfields.EndpointState + ".from": fromState,
 			logfields.EndpointState + ".to":   toState,
-			"file": fileName,
-			"line": fileLine,
+			"file":                            fileName,
```
Comment (ianvernon) [inline pkg/endpoint/endpoint.go]: I had `go fmt` do this in go 1.11, but the CI has a different version of `go fmt`, which causes the formatting precheck to fail. You may want to run `gofmt` of a specific version until we update the CI :/

**Review by ianvernon — APPROVED** [review]

File: pkg/endpoint/endpoint.go (hunk 4 above)
Comment (ianvernon) [reply pkg/endpoint/endpoint.go]: Never mind, the build already got past that point. I stand corrected!

**Comment by aanm** [comment]
test-me-please

**Comment by tgraf** [comment]
@aanm Supportive of the changes, can we wait until #5358 is merged?

File: pkg/endpoint/directory.go [hunk 5]
```diff
@@ -17,37 +17,44 @@ package endpoint
...
 	scopedLog := e.getLogger()
 
-	tmpDir := getTempEndpointDirectory(origDir)
+	tmpDir := e.NextDirectoryPath()
 	// If generation failed, keep the directory around. If it ever succeeds
 	// again, clean up the XXX_next_fail copy.
 	failDir := e.FailedDirectoryPath()
 	os.RemoveAll(failDir) // Most likely will not exist; ignore failure.
```
Comment (joestringer) [inline pkg/endpoint/directory.go:57]: I don't think that this line (the remove of any potentially existing failDir) will be necessary any more.

**Comment by raybejjani** [comment]
@aanm I'm afraid you lost a merge race somewhere. The PR now has conflicts, a terrible PR disease...

**Review by tgraf — CHANGES_REQUESTED** [review]
We will wait for #5358 to be merged

**Review by tgraf — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Pull request merged**